rlu/src/stm/version.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! The version lock is a special type word sized spin lock, that
+//! contains a single bit to indicate a lock, while using the rest
+//! of the bits for versioning.
+
+use crate::stm::error::*;
+use std::{
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+/// A [`VersionLock`] is a combination of a simple bounded spin-locking mechanism, needing
+/// 1-bit of a word-sized value to lock a certain region. The rest of the value is being
+/// used to increment a version counter. Use a [`VersionClock`], when you want to
+/// implement versioned updates on transactional memory regions.
+#[derive(Default, Clone, Debug)]
+pub struct VersionLock {
+    atomic: Arc<AtomicUsize>,
+}
+
+impl VersionLock {
+    /// Creates a new [`VersionLock`] with the desired version
+    pub fn new(version: usize) -> Self {
+        Self {
+            atomic: Arc::new(AtomicUsize::new(version)),
+        }
+    }
+
+    /// Tries to acquire a lock and returns an `Ok(())` on success.
+    ///
+    /// # Example
+    /// ```
+    /// use stronghold_rlu::stm::VersionLock;
+    /// let lock = VersionLock::default();
+    /// lock.try_lock().expect("Failed to acquire lock");
+    /// assert!(lock.is_locked());
+    /// ```
+    pub fn try_lock(&self) -> Result<(), TxError> {
```

Comment:
This spinlock is not bounded. My guess is that we can have a deadlock when two threads try to lock the tvar of their write set at the same time. That's why the paper asks for bounded spinlock

Reply:
True. The question is, if a bounded spinlock needs some empirical value to actually make the algorithm work. The bounded version is currently commented out, but would offer a very high bound of runs. Using sleep reduces the CPU load.